### PR TITLE
Add manual journal entry form

### LINKED ISSUE #2144: Journal tab raises UnicodeDecodeError when filtering test contains umlaut
https://sentry.4teamwork.ch/sentry/onegov-gever/issues/1238/

```
UnicodeDecodeError: 'ascii' codec can't decode byte 0xc3 in position 3: ordinal not in range(128)
(5 additional frame(s) were not displayed)
...
  File "home/zope/theta.4teamwork.ch/17-lab.onegovgever.ch/src/ftw.tabbedview/ftw/tabbedview/browser/listing.py", line 194, in update
    query = self.table_source.build_query()
  File "home/zope/theta.4teamwork.ch/17-lab.onegovgever.ch/src/opengever.core/opengever/tabbedview/filtered_source.py", line 7, in build_query
    query = super(FilteredTableSourceMixin, self).build_query()
  File "home/zope/eggs/ftw.table-1.17.0-py2.7.egg/ftw/table/basesource.py", line 68, in build_query
    query, self.config.filter_text)
  File "home/zope/theta.4teamwork.ch/17-lab.onegovgever.ch/src/opengever.core/opengever/journal/tab.py", line 129, in extend_query_with_textfilter
    results = filter(_search_method, results)
  File "home/zope/theta.4teamwork.ch/17-lab.onegovgever.ch/src/opengever.core/opengever/journal/tab.py", line 116, in _search_method
    if text.lower() in title_helper(item, '').lower():
```

### PR DESCRIPTION
**Pull request by phgross**
bildschirmfoto 2016-09-19 um 12 30 05

bildschirmfoto 2016-09-19 um 12 28 56

See [Todo](https://basecamp.com/2768704/projects/12330648/todos/271609550) for more details

Besides the new functionality to add a journal entry manually, this PR fixes #2144 and changes the wrapper id for contact objects, it uses now `contact-{id}` instead of `person-{id}` or `organization-{id}`.

The styling adjustments are in separate [PR in plonetheme.teamraum](https://github.com/4teamwork/plonetheme.teamraum/pull/464).

@deiferni @lukasgraf

### CODE REVIEW

File: opengever/contact/contactfolder.py [hunk 1]
```diff
@@ -37,17 +37,14 @@ def _getOb(self, id_, default=_marker):
...
-        if id_.startswith('organization-'):
-            organization_id = int(id_.split('-')[-1])
-            organization = Organization.query.get(organization_id)
-            if organization:
-                return OrganizationWrapper.wrap(self, organization)
+        if id_.startswith('contact-'):
+            contact_id = int(id_.split('-')[-1])
+            contact = Contact.query.get(contact_id)
+            if contact:
+                if isinstance(contact, Person):
```
Comment (deiferni) [inline opengever/contact/contactfolder.py]: it should be a warning sign if you have to do an `isinstance` check. usually there is a more object oriented way to achieve your goal. 

here you could reverse the calls and ask the object for its wrapped version.

File: opengever/journal/entry.py [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+from persistent.list import PersistentList
+from plone import api
+from zope.component import getUtility
+from zope.event import notify
+from zope.globalrequest import getRequest
+from zope.i18n import translate
+from zope.schema.interfaces import IVocabularyFactory
+
+
+MANUAL_JOURNAL_ENTRY = 'manually journal entry'
```
Comment (deiferni) [inline opengever/journal/entry.py]: This seems to be some kind of id for the entry type?
I would tend towards not using whitespaces in such a string. Also i'd adapt the naming a little:

```python
MANUAL_JOURNAL_ENTRY_TYPE = 'manual_journal_entry'
```

File: opengever/journal/entry.py [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+
+
+class ManuallJournalEntry(object):
```
Comment (deiferni) [inline opengever/journal/entry.py]: typo

Comment (deiferni) [inline opengever/journal/entry.py]: missing docstring (purpose of class)

File: opengever/journal/entry.py [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+        notify(JournalEntryEvent(**entry))
+
+    def serialize_documents(self):
+        """Returns a persistent list of dicts with the following data for for
+        all documents:
+
+        id: oguid
+        title: document title
+        """
+
```
Comment (deiferni) [inline opengever/journal/entry.py]: ✂️

File: opengever/journal/entry.py [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+                {'id': Oguid.for_object(doc).id, 'title': doc.title}))
+
+        return value
+
+    def serialize_contacts(self):
+        """Returns a persistent list of dicts for all contacts.
+        """
+        value = PersistentList()
+        for item in self.contacts:
+            if isinstance(item, OrgRole):
```
Comment (deiferni) [inline opengever/journal/entry.py]: nah-ah.

File: opengever/journal/entry.py [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+                                    'title': item.get_title()}))
+            else:
+                value.append(
+                    PersistentDict({'id': item.contact_id,
+                                    'title': item.get_title()}))
+
+        return value
+
+    def get_title(self):
+        msg = _(u'label_manual_entry', default=u'Manual entry: ${category}',
```
Comment (deiferni) [inline opengever/journal/entry.py]: label_manual_journal_entry

File: opengever/journal/entry.py [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+                mapping={'category': self.get_category_title()})
+
+        return translate(msg, context=getRequest())
+
+    def get_category_title(self):
+        voca_factory = getUtility(
+            IVocabularyFactory,
+            name='opengever.jounral.manual_entry_categories')
```
Comment (deiferni) [inline opengever/journal/entry.py]: `jounral` 😲

File: opengever/journal/form.py [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+from plone.directives import form
+from plone.formwidget.autocomplete import AutocompleteMultiFieldWidget
+from z3c.form.field import Fields
+from z3c.form.form import AddForm
+from z3c.relationfield.schema import RelationChoice
+from z3c.relationfield.schema import RelationList
+from zope import schema
+
+
+class IManaulJournalEntry(form.Schema):
```
Comment (deiferni) [inline opengever/journal/form.py]: `Manaul` - if i find another one i get a free beer 🍻

File: opengever/journal/form.py [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+        default=[],
+        missing_value=[],
+        value_type=RelationChoice(
+            title=u"Related",
+            source=RepositoryPathSourceBinder(
+                portal_type=("opengever.document.document", "ftw.mail.mail"),
+                navigation_tree_query={
+                    'object_provides':
+                    ['opengever.repository.repositoryroot.IRepositoryRoot',
+                     'opengever.repository.repositoryfolder.' +
```
Comment (deiferni) [inline opengever/journal/form.py]: i would avoid the break here for the sake of readability

File: opengever/journal/form.py [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+        title=_(u'label_related_documents', default=u'Related Documents'),
+        default=[],
+        missing_value=[],
+        value_type=RelationChoice(
+            title=u"Related",
+            source=RepositoryPathSourceBinder(
+                portal_type=("opengever.document.document", "ftw.mail.mail"),
+                navigation_tree_query={
+                    'object_provides':
+                    ['opengever.repository.repositoryroot.IRepositoryRoot',
```
Comment (deiferni) [inline opengever/journal/form.py]: do we really want the repo root in here?

File: opengever/journal/form.py [hunk 11]
```diff
@@ -0,0 +1,72 @@
...
+
+    fields['contacts'].widgetFactory = AutocompleteMultiFieldWidget
+
+    def createAndAdd(self, data):
+        entry = ManuallJournalEntry(self.context,
+                                    data.get('category'),
+                                    data.get('comment'),
+                                    data.get('contacts'),
+                                    data.get('related_documents'))
+
```
Comment (deiferni) [inline opengever/journal/form.py]: ✂️

File: opengever/journal/form.py [hunk 12]
```diff
@@ -0,0 +1,72 @@
...
+                     'opengever.dossier.behaviors.dossier.IDossierMarker',
+                     'opengever.document.document.IDocumentSchema',
+                     'ftw.mail.mail.IMail', ]
+                }),
+            ),
+        required=False,
+    )
+
+
+class ManualJournalEntryForm(AddForm):
```
Comment (deiferni) [inline opengever/journal/form.py]: `ManualJournalEntryAddForm`? There may be an edit-form, too some day.

File: opengever/journal/locales/de/LC_MESSAGES/opengever.journal.po [hunk 13]
```diff
@@ -20,20 +20,42 @@ msgid "heading_journal"
...
-#: ./opengever/journal/tab.py:68
+#: ./opengever/journal/tab.py:74
 msgid "label_actor"
 msgstr "Geändert von"
 
+#. Default: "Add journal entry"
+#: ./opengever/journal/form.py:56
+#: ./opengever/journal/templates/journal_selection.pt:2
+msgid "label_add_journal_entry"
+msgstr "Journal Eintrag hinzufügen"
```
Comment (deiferni) [inline opengever/journal/locales/de/LC_MESSAGES/opengever.journal.po]: Journaleintrag

File: opengever/journal/tab.py [hunk 14]
```diff
@@ -113,7 +138,7 @@ def extend_query_with_textfilter(self, results, text):
 
         def _search_method(item):
             # title
-            if text.lower() in title_helper(item, '').lower():
+            if text.lower() in title_helper(item, u'').lower().encode('utf-8'):
```
Comment (deiferni) [inline opengever/journal/tab.py]: we should use unicode internally, so `text.decode('utf-8')` would be better

File: opengever/journal/templates/journal_references.pt [hunk 15]
```diff
@@ -0,0 +1,20 @@
...
+    </li>
+  </ul>
+</div>
+
+<div class="documents" tal:condition="view/documents">
+  <p i18n:domain="opengever.journal" i18n:translate="label_documents">Documents</p>
+  <ul class="documents">
+    <li tal:repeat="document view/documents">
+      <a href="#"
+         tal:attributes="href string: ${context/absolute_url}/resolve_oguid?oguid=${document/id}"
```
Comment (deiferni) [inline opengever/journal/templates/journal_references.pt]: i'd use the view-prefix`@@resolve_oguid`

File: opengever/journal/tests/test_journal_tab.py [hunk 16]
```diff
@@ -0,0 +1,67 @@
...
+
+
+class TestJournalTab(FunctionalTestCase):
+
+    def setUp(self):
+        super(TestJournalTab, self).setUp()
+
+    @browsing
+    def test_list_all_journal_entries(self, browser):
+
```
Comment (deiferni) [inline opengever/journal/tests/test_journal_tab.py]: ✂️

File: opengever/journal/tests/test_manual_entry.py [hunk 17]
```diff
@@ -0,0 +1,93 @@
...
+class TestManualJournalEntry(FunctionalTestCase):
+
+    def setUp(self):
+        super(TestManualJournalEntry, self).setUp()
+        self.dossier = create(Builder('dossier'))
+        self.contactfolder = create(Builder('contactfolder'))
+
+    @browsing
+    def test_adds_entry_to_currents_context_journal(self, browser):
+
```
Comment (deiferni) [inline opengever/journal/tests/test_manual_entry.py]: ✂️

Comment (deiferni) [inline opengever/journal/tests/test_manual_entry.py]: current

File: opengever/journal/form.py (hunk 10 above)
Comment (phgross) [inline opengever/journal/form.py]: Yes it should be possible to reference a document from an other repository root, the `object_provides` in the navigation_tree_query only limits the objects which can be use for navigation not the objects which actually can be selected as a reference.

File: opengever/journal/form.py (hunk 12 above)
Comment (phgross) [inline opengever/journal/form.py]: No! 😉 

But I'll update the name.

Comment (phgross) [inline opengever/journal/form.py]: No! 😉 

But I'll update the name.

File: opengever/journal/tab.py (hunk 14 above)
Comment (phgross) [inline opengever/journal/tab.py]: I know, but there are already "a few" journal entries in production so to migration of all them is not a option right now. Should be solved when chaning to ftw.activity.

**Comment by phgross** [comment]
@deiferni updated please have another look ...

**Comment by phgross** [comment]
@deiferni updated again ...

**Pull request merged**